### DISCUSSION TITLE
Adding a new executable format to mold

### DISCUSSION THREAD

**Discussion by Luminoso-256** [rui314/mold#346] [General]
Hey there!

If I wanted to add a new executable format as an output type for mold's linking, where would I start?

**Comment by rui314** [comment]
It depends on what you are trying. If your file format is a small extension to ELF, it shouldn't be too hard to support it. If you want to define something completely new, you essentially have to write a linker from scratch.

**Reply by BenBE** [reply]
Do I read this correctly as "supporting Portable Executables (AKA MZ/PE) for Windows targets will require a near complete rewrite of mold"?

**Reply by rui314** [reply]
Yes. And that's what I'm doing for mold/macOS. mold shares the design of the code but don't share code itself (except a few files in the top directory). lld is actually written that way and I think it was a correct design choice.

**Comment by rui314** [comment]
I didn't answer to "where to start" part. I think if your want to write a linker for a new file format, you should start writing a file dumper to familiarize yourself with the new file format. Producing a file for an unfamiliar file format is very hard, so it needs a training period to verify that you're knowledge of the format is correct. That's what I did first when I started writing mold/macOS.

**Reply by Luminoso-256** [reply]
Alright. I have written a dumper for the format (it's a simple format I came up with myself)
